core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java

```diff
@@ -0,0 +1,159 @@
+import ch.ethz.ssh2.Session;
+import ch.ethz.ssh2.StreamGobbler;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+
+import co.cask.cdap.etl.api.action.Action;
+import co.cask.cdap.etl.api.action.ActionContext;
+import com.google.common.base.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import javax.annotation.Nullable;
+
+/**
+ * SSH into a remote machine and execute a script on that machine.
+ * A user must specify username and keypair authentication credentials.
+ * Optionals include port and machine URL
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("SSHAction")
+@Description("Action to run a script on remote machine")
+public class SSHAction extends Action {
+  private final SSHActionConfig config;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SSHAction.class);
+
+  public SSHAction(SSHActionConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(final ActionContext context) throws Exception {
+    Connection connection = new Connection(config.host);
+    try {
+      connection.connect();
+
+      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
+      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+        throw new IOException(String.format("Unable to establish SSH connection for %s@%s on port %d",
+                                            config.user, config.host, config.port));
+      }
+
+      LOG.info(String.format("Connection established with the host %s", config.host));
+
+      Session session = connection.openSession();
+      session.execCommand(config.cmd);
+
+      InputStream stdout = new StreamGobbler(session.getStdout());
+      BufferedReader outBuffer = new BufferedReader(new InputStreamReader(stdout));
+      InputStream stderr = new StreamGobbler(session.getStderr());
+      BufferedReader errBuffer = new BufferedReader(new InputStreamReader(stderr));
+
+      StringBuilder outBuilder = new StringBuilder();
+      StringBuilder errBuilder = new StringBuilder();
+
+      String line = outBuffer.readLine();
+      while (line != null) {
+        outBuilder.append(line).append("\n");
+        line = outBuffer.readLine();
+      }
+
+      line = errBuffer.readLine();
+      while (line != null) {
+        errBuilder.append(line).append("\n");
+        line = errBuffer.readLine();
+      }
+
+      LOG.info("Output: {}", outBuilder.toString());
+      LOG.info("Errors: {}", errBuilder.toString());
+
+      if (config.output != null) {
+        context.getArguments().set(config.output, outBuilder.toString());
+      }
+      session.close();
```

Comment:
Do this need to be in finally? Or the connection.close() will close this as well?